test/unittests/utils/test_connection_live.py

```diff
@@ -1,4 +1,217 @@
+
+
+def test_put() -> None:
+    con = ConnectionLive("http://example.com/")
+    con._try_network_action = Mock()
+    con.put(route="/v2/values")
+    expected_params: RequestParameters = con._try_network_action.call_args.args[0]
+    assert expected_params.method == "PUT"
+    assert expected_params.url == "http://example.com/v2/values"
+    assert expected_params.data is None
+    assert expected_params.headers is None
+    assert expected_params.files is None
+
+
+def test_put_with_data() -> None:
+    con = ConnectionLive("http://example.com/")
+    con._try_network_action = Mock()
+    con.put(route="/v2/values", data={"foo": "bar"})
+    expected_params: RequestParameters = con._try_network_action.call_args.args[0]
+    assert expected_params.method == "PUT"
+    assert expected_params.url == "http://example.com/v2/values"
+    assert expected_params.data == {"foo": "bar"}
+    assert expected_params.headers == {"Content-Type": "application/json; charset=UTF-8"}
+    assert expected_params.files is None
+
+
+def test_delete() -> None:
+    con = ConnectionLive("http://example.com/")
+    con._try_network_action = Mock()
+    con.delete(route="/v2/values")
+    expected_params: RequestParameters = con._try_network_action.call_args.args[0]
+    assert expected_params.method == "DELETE"
+    assert expected_params.url == "http://example.com/v2/values"
+    assert expected_params.data is None
+    assert expected_params.headers is None
+    assert expected_params.files is None
+
+
+def test_default_timeout() -> None:
+    con = ConnectionLive("http://example.com/")
+    con._try_network_action = Mock()
+    for method in (con.post, con.put):
+        method = cast(Callable[..., Any], method)
+        method(route="/v2/resources")
+        expected_params: RequestParameters = con._try_network_action.call_args.args[0]
+        assert expected_params.timeout == con.timeout_put_post
+    for method in (con.get, con.delete):
+        method = cast(Callable[..., Any], method)
+        method(route="/v2/resources")
+        expected_params = con._try_network_action.call_args.args[0]
+        assert expected_params.timeout == con.timeout_get_delete
+
+
+def test_custom_timeout() -> None:
+    con = ConnectionLive("http://example.com/")
+    con._try_network_action = Mock()
+    con.post(route="/v2/resources", timeout=1)
+    expected_params: RequestParameters = con._try_network_action.call_args.args[0]
+    assert expected_params.timeout == 1
+
+
+def test_custom_header() -> None:
+    con = ConnectionLive("http://example.com/")
+    con._try_network_action = Mock()
+    for method in (con.post, con.put, con.get, con.delete):
+        method = cast(Callable[..., Any], method)
+        method(route="/v2/resources", headers={"foo": "bar"})
+        expected_params: RequestParameters = con._try_network_action.call_args.args[0]
+        assert expected_params.headers == {"foo": "bar"}
+
+
+def test_custom_content_type() -> None:
```

Comment:
It would mean a lot more writing, but in my opinion, it would give more readable results if you created a class for the functions on 177, 187, 197, 207 and then create a function  for the lines that you call within the loop (eg. 181-184) and for each method you call this mini function wrapped in a separate function:

```
def helper_with_assert(con) -> None:
   con.post(route="/v2/resources", headers={"Content-Type": "bar"})
        expected_params: RequestParameters = con._try_network_action.call_args.args[0]
        assert expected_params.headers == {"Content-Type": "bar"}

def con_post() -> None:
       con = ConnectionLive("http://example.com")
       con._try_network_action = Mock()
      helper_with_assert(con)
```

In case of an error, the result would be a lot cleaner and the function that has the error would be easier to identify.

Reply:
Luckily, I could achieve a better feedback message with a simple trick: 

```
for method in (con.post, con.put, con.get, con.delete):
        assert x = y, f"Method '{method.__name__}' failed"
```